src/database/postgres.go

```diff
@@ -0,0 +1,198 @@
+    defer tx.Rollback(ctx)
+
+    batch := &pgx.Batch{}
+    for _, entry := range index.Entries {
+        wordOccurrencesJSON, err := json.Marshal(entry.WordOccurrences)
+        if err != nil {
+            return fmt.Errorf("failed to marshal word occurrences: %w", err)
+        }
+
+        batch.Queue(`
+            INSERT INTO index_list(
+                name, uri, data_type, source_type, first_added_time, 
+                last_refreshed_time, content, search_vector, word_occurrences
+            )
+            VALUES($1, $2, $3, $4, $5, $6, $7, to_tsvector('english', $7), $8)
+            ON CONFLICT (name) DO UPDATE SET
+                uri = EXCLUDED.uri,
+                data_type = EXCLUDED.data_type,
+                source_type = EXCLUDED.source_type,
+                last_refreshed_time = EXCLUDED.last_refreshed_time,
+                content = EXCLUDED.content,
+                search_vector = to_tsvector('english', EXCLUDED.content),
+                word_occurrences = EXCLUDED.word_occurrences
+        `, entry.Name, entry.URI, entry.DataType.String(), entry.SourceType.String(),
+            entry.FirstAddedTime.AsTime(), entry.LastRefreshedTime.AsTime(),
+            entry.Content, wordOccurrencesJSON)
+    }
+
+    br := tx.SendBatch(ctx, batch)
+    _, err = br.Exec()
+    br.Close()
+    if err != nil {
+        return fmt.Errorf("failed to insert rows: %w", err)
+    }
+
+    err = tx.Commit(ctx)
+    if err != nil {
+        return fmt.Errorf("failed to commit transaction: %w", err)
+    }
+
+    // Refresh the materialized view outside the transaction
+    _, err = conn.Exec(ctx, `REFRESH MATERIALIZED VIEW CONCURRENTLY mv_index_list`)
+    if err != nil {
+        return fmt.Errorf("failed to refresh materialized view: %w", err)
+    }
+
+    return nil
+}
+
+
+// queryRow retrieves a single row from the index_list table based on the provided name.
+// It returns a pointer to a pb.IndexListEntry struct containing the row data, or an error if the query fails.
+//
+// Parameters:
+//   - ctx: The context for database operations.
+//   - conn: A PgxIface interface for database connection.
+//   - name: The name of the index entry to retrieve.
+//
+// Returns:
+//   - *pb.IndexListEntry: A pointer to the retrieved index entry.
+func queryRow(ctx context.Context, conn PgxIface, name string) (*pb.IndexListEntry, error) {
+	tx, err := conn.Begin(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to database: %w", err)
+	}
+	defer tx.Rollback(ctx)
+
+	var entry pb.IndexListEntry
+	var dataType, sourceType string
+	var firstAddedTime, lastRefreshedTime time.Time
+	var wordOccurrencesJSON []byte
+
+	err = tx.QueryRow(ctx, `
+		SELECT name, uri, data_type, source_type, first_added_time, last_refreshed_time, content, word_occurrences
+		FROM index_list 
+		WHERE name=$1
+	`, name).Scan(
+		&entry.Name, &entry.URI, &dataType, &sourceType, &firstAddedTime, &lastRefreshedTime, &entry.Content, &wordOccurrencesJSON)
```

Comment:
If we used a JSONB representation of the proto this would be so much simpler

Reply:
Yes done